1389: Refactor categorized sort order to use DB column instead of site properties XML

payten opened the pull request:
Delivers #1389.

Hi @steveswinsburg.  Apologies for the massive PR, but a fair bit went into the refactor.  It could all use a review!  Happy to talk it through on a call if preferred. 

This PR contains 3 commits:

1. the refactor of the assignments sorting within categories (moved from the site properties XML to a new table column).  I've added the new column to the Sakai 11 upgrade script.  Does it need to be somewhere else to be included in the auto.ddl upgrade?

2. new feature to allow sorting of categories within the settings.  This makes use of the existing category_order column in gb_category_t.  I've tried to retain the service API so old Gradebook should still be happy.

3. When saving Settings, set the response page to a new instance of SettingsPage so all components are re-rendered to reflect any model changes.  We retained the page instance previously to preserve the state of the accordion, but I wasn't able to get the category panel to re-render the table when the sort order had changed.  I've instead added an `expanded` boolean to those panels so we can programmatically set the state of the accordion on a new instance of SettingsPage.

steveswinsburg commented:
Hey @payten, I am looking at this, just need to finish the review :)

steveswinsburg commented on edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java:
```diff
@@ -3269,5 +3290,93 @@ public Object doInHibernate(Session session) throws HibernateException {
+		}
+
+		if(order == null) {
+			throw new IllegalArgumentException("Categoryized Order cannot be null");
```
typo 'Categoryized'

steveswinsburg replied:
Man I hate the US Spelling with a 'z', but since this is in the service :(

steveswinsburg commented on edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java:
```diff
@@ -3113,6 +3117,19 @@ public void updateGradebookSettings(String gradebookUid, GradebookInformation gb
 		//If category does not have an ID it is new
 		//If category has an ID it is to be updated. Update and remove from currentCategoryMap.
 		//Any categories remaining in currentCategoryMap are to be removed.
+		Collections.sort(newCategoryDefinitions, new Comparator<CategoryDefinition>() {
```
Could you stick an additional doc line in here just saying that we need to sort them first so that new ones get created in the correct order?

steveswinsburg replied:
Also, could this use CategoryOrderComparator ?

steveswinsburg commented on edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java:
```diff
@@ -3269,5 +3290,93 @@ public Object doInHibernate(Session session) throws HibernateException {
+		assignmentsInNewCategory.add(null); //ensure size remains the same for the remove
+		assignmentsInNewCategory.remove(target); //remove item
+		assignmentsInNewCategory.add(order, target); //add at ordered position, will shuffle others along
+
```
Good work reusing this stuff from the GradebookNgBusinessService. Took me a while to figure out this shuffle approach, had it all sketched out on paper, ha!

payten replied:
You do the thinking so I don't have to ;)

steveswinsburg commented on gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/pages/GradebookPage.java:
```diff
@@ -155,6 +157,11 @@ public boolean isVisible() {
 
+		if (settings.isCategoriesEnabled()) {
+			// pre-sort assignments by the categorized sort order
+			Collections.sort(assignments, new CategorizedAssignmentComparator());
```
I wonder if this should go in the backend and be passed into the matrix? The list of assignments is already sorted in getGradebookAssignments so this will sort twice but maybe we can't avoid that and it's not a big deal.

However note that students don't have access to the GradebookInformation so it would need to be wrapped in a role check if added to the backend.

Probably a lot of work when it could just be left here. Maybe just document why it is here and we are done.

payten replied:
Yeah, I think we could move this to the service to by extending the API of `getViewableAssignmentsForCurrentUser` to define a SortType and then by introducing a new SortType.SORT_BY_CATEGORY. We could then offering our fancy comparator in the service as a static on `Assignment` which would be used when SORT_BY_CATEGORY was chosen.  I've been hesitant to change too much in the service due to the fear of breaking Gradebook or Gradebook2!

Happy to have a go at the refactor if you think it's worth it?

steveswinsburg commented on gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/pages/GradebookPage.java:
```diff
@@ -648,4 +646,46 @@ private Label constructTableLabel(final String componentId, final DataTable tabl
+	/**
+	 * Comparator class for sorting Assignments in their categorised ordering
+	 */
+	class CategorizedAssignmentComparator implements Comparator<Assignment> {
```
We have a bunch of comparators in the service, and now one here, maybe these should be split out into the business.util package?

steveswinsburg commented on gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/AssignmentColumnHeaderPanel.java:
```diff
@@ -337,4 +343,30 @@ private Component getParentCellFor(final Component component) {
+			// if no categorized order for assignment, calculate one based on the default sort order
+			List<Assignment> assignments = AssignmentColumnHeaderPanel.this.businessService.getGradebookAssignments();
+			List<Long> assignmentIdsInCategory = assignments.stream()
+																												.filter(a -> a.getCategoryId() == assignment.getCategoryId())
```
weird spacing here, baby on the tab key?

payten commented:
Thanks for the review @steveswinsburg!  I'll work through your comments and poke you when I've fixed up the issues.

payten commented:
Hi @steveswinsburg, I've just pushed some refactorings based on your comments.  Namely these things:
- fix Categoryized type
- move CategoryOrderComparator to CategoryDefinition.orderComparator so it can be used by both the service and tool
- fix cray cray indenting
- move the assignments comparator (for sorting by category or order within category) from the tool to the service and created a new SortType so the sorting is only performed once by the service and doesn't need to be resorted in the tool. Also changed the Categories button to refresh the entire page to ensure the sort order is updated from the service (rather than in the javascript).
- also fixed an issue whereby Assignment#sort_order was not being returned in the SORT_BY_SORTING order due to use of a TreeSet (changed to LinkedHashSet - see #1536)
